Update riscv isa

patrick-rivos opened the pull request:
The RISCV ISA is roughly 5 years outdated at this point. Existing instructions still work great, but there have been new extensions released in the meantime. One of those extensions is the Vector extension, which encodes a fair amount of complexity with its variable length instruction encoding. [Vector spec](https://github.com/riscv/riscv-v-spec/blob/master/v-spec.adoc#risc-v-v-vector-extension)

This PR adds a parser for the RISCV instruction set and references binutils 2.40 as a source of truth. This PR is a **draft** and still has work  to be done (memory reads/writes using vector, actually encoding the binary format, etc.)

rbertran commented on .gitmodules:
I assume that the parser only requires 1 or 2 files from binutils source tree. Instead of adding the entire binutils tree as a dependency, I'd prefer to just add the files required (or a script to get them).

patrick-rivos replied:
Yep - it only parses 2 files iirc.
Adding a submodule doesn't clone in the entire source unless you do `git submodule update --init`. On normal clones/usages its super lightweight (just this txtfile). Submodules make it explicitly clear which source hash was used (and how to easily get it).
If you'd still prefer to just copy the files over just LMK and I'm happy to do that instead.

rbertran replied:
I think that adding as a submodule creates the impression of "strong" dependency... I'd prefer to just copy the files you need and update them when there are new binutils releases, or a better alternatively would be to add a simple shell script to fetch the files from the latest (or specific) binutils release. At the end of the day, these are "development mode" files that will only be needed once to create the YAML files (which will be committed into the repo).

rbertran commented on targets/riscv/isa/riscv-common/instruction.yaml:
I know that is not documented anywhere... but the conventions on the YAML instruction entries are the following:

```
- Name: "C.SD_V0"  -> Instruction name in capital letters. Usually is the assembly mnemonic
                                     plus "_V#" that indicates the instruction variant starting with 0.
                                     Sometimes the same instruction/mnemonic is implemented differently
                                     in the microrarchitecture depending on the input parameters
                                     (e.g. add immeidate can have two implementations: the regular
                                     one and when the immediate
                                     is 0 , which is essentially a no-operation)
  Mnemonic: "C.SD" -> mnemonic (string that will be generated) in capital letters
  Opcode: "0" -> opcode
  Format: "cl_d" -> instruction format
  Operands:                             -> Default operands are set from the instruction format, but if one
     funct3: ['7', 'funct3', '?']           need to override them you can do so with this entries. Check existing
                                                    RISCV definitions or the POWER backend for examples.
  MemoryOperands:  -> Memory operands are defined here.
    ....
  ImplicitOperands:
```
Make sure your parser generates maintaining this format and conventions.

patrick-rivos replied:
This is super helpful - Thank you!

rbertran commented:
Thanks, I added some feedback on some files. I assume that this is still a WIP PR, once it is more mature and passes the CI I'll integrate it. From the CI output, it looks like something broke the existing code generation fon RISCV.

If you hit a roadblock, let me know and I'll try to debug what can be going on.

patrick-rivos commented:
Thanks! I'm pretty sure the CI is failing since we generate incorrect binary representations for insns.

And yes - it's an early WIP PR to make sure my general approach will be mergable once it's ready :)

patrick-rivos commented:
This is still a draft since I haven't had the time to dig into the actual opcodes - It's functional for generating assembly.

I updated the format/naming conventions to match existing instructions (V0 is unmasked operations or masked with no unmasked equivalent, V1 is masked). Also removed the submodule dependency on binutils. I'm not sure when I'll have the time to get the vector opcodes complete, but when I do I'll chip away at it and keep this PR up-to-date with my progress.

patrick-rivos commented:
Rebased and added some asm-level features on top.
At a high level still where it was before - generates valid assembly but the opcodes are TODO.

I do have a question about target-specific flags in generic tools:
In `RISC-V: Expose sew and lmul to mp_seq.py command line` I make -lmul and -sew settable on the command line but they have no meaning for the POWERPC target. Should I think of another way to expose those toggles or is this approach OK?

rbertran commented:
Thanks @patrick-rivos for your contributions. Regarding your question, there is not a way to just show the arguments when a particular target is selected. For now, I think that best option to handle _**target specific**_ command line arguments is to clearly specify the specific target in the argument description. I.e. `"Selected element width for vector insns"` becomes  `"Selected element width for vector insns (only valid for RISCV backend)"` . Also, I'd group such options within another target specific group. Use `cmdline.add_group` to add another group with the description `"RISCV Specific options"` or similar.
